libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java

```diff
@@ -185,29 +187,38 @@ private static boolean isServerModule(Module requestingModule) {
         return requestingModule.isNamed() && requestingModule.getLayer() == ModuleLayer.boot();
     }
 
-    private static Module requestingModule(Class<?> callerClass) {
+    static Module requestingModule(Class<?> callerClass) {
         if (callerClass != null) {
             Module callerModule = callerClass.getModule();
             if (systemModules.contains(callerModule) == false) {
                 // fast path
                 return callerModule;
             }
         }
-        int framesToSkip = 1  // getCallingClass (this method)
-            + 1  // the checkXxx method
-            + 1  // the runtime config method
-            + 1  // the instrumented method
-        ;
         Optional<Module> module = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE)
-            .walk(
-                s -> s.skip(framesToSkip)
-                    .map(f -> f.getDeclaringClass().getModule())
-                    .filter(m -> systemModules.contains(m) == false)
-                    .findFirst()
-            );
+            .walk(s -> findRequestingModule(s.map(f -> moduleOf(f.getDeclaringClass()))));
         return module.orElse(null);
     }
 
+    /**
+     * @throws NullPointerException if the requesting module is {@code null}
+     */
+    static Optional<Module> findRequestingModule(Stream<Module> modules) {
+        return modules.map(Objects::requireNonNull)
+            .dropWhile(not(systemModules::contains)) // Skip the entitlements runtime
```

Comment:
could this more explicitly check the entitlements runtime, rather than assuming anything not in system modules is entitlements?

Reply:
Yeah I'd definitely rather do that. Do you have any suggestions?

Reply:
I think Jack's suggestion below (a fixed set of classes from entitlements, or alternatively, the entitlement module) could work.

Reply:
Ok the challenge there is, we're filtering on modules rather than classes. Since we run the unit tests non-modular, we can't expect the module info on the entitlement classes to match what they will be in production.

Let me see if I can find something clean here. We'll want to skip some frames based on class and others based on module, and do so in a way that still works in unit tests. 🤔

Reply:
It's fine, it turned out the code is actually a little simpler this way: [a780d69](https://github.com/elastic/elasticsearch/pull/118983/commits/a780d695a6e929b1c2de4aaa27ccd5ff5b65ce44)